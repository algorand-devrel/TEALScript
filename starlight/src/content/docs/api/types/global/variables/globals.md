---
editUrl: false
next: false
prev: false
title: "globals"
---

> **`const`** **globals**: `Object`

## Type declaration

### assetCreateMinBalance

> **assetCreateMinBalance**: [`uint64`](../type-aliases/uint64.md)

### assetOptInMinBalance

> **assetOptInMinBalance**: [`uint64`](../type-aliases/uint64.md)

### callerApplicationAddress

> **callerApplicationAddress**: [`Address`](../classes/Address.md)

### callerApplicationID

> **callerApplicationID**: [`Application`](../classes/Application.md)

### creatorAddress

> **creatorAddress**: [`Address`](../classes/Address.md)

### currentApplicationAddress

> **currentApplicationAddress**: [`Address`](../classes/Address.md)

### currentApplicationID

> **currentApplicationID**: [`Application`](../classes/Application.md)

### genesisHash

> **genesisHash**: [`bytes32`](../type-aliases/bytes32.md)

### groupID

> **groupID**: [`bytes32`](../type-aliases/bytes32.md)

### groupSize

> **groupSize**: [`uint64`](../type-aliases/uint64.md)

### latestTimestamp

> **latestTimestamp**: [`uint64`](../type-aliases/uint64.md)

### logicSigVersion

> **logicSigVersion**: [`uint64`](../type-aliases/uint64.md)

### maxTxnLife

> **maxTxnLife**: [`uint64`](../type-aliases/uint64.md)

### minBalance

> **minBalance**: [`uint64`](../type-aliases/uint64.md)

### minTxnFee

> **minTxnFee**: [`uint64`](../type-aliases/uint64.md)

### opcodeBudget

> **opcodeBudget**: [`uint64`](../type-aliases/uint64.md)

### round

> **round**: [`uint64`](../type-aliases/uint64.md)

### zeroAddress

> **zeroAddress**: [`Address`](../classes/Address.md)

## Source

[types/global.d.ts:800](https://github.com/algorandfoundation/tealscript/blob/18ba30a9/types/global.d.ts#L800)
