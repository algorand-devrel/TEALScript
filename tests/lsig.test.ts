/* eslint-disable func-names */
/* eslint-disable prefer-arrow-callback */
import { describe, expect, test } from '@jest/globals';
import algosdk from 'algosdk';
import { artifactsTest, compileAndCreate } from './common';

const PATH = 'tests/contracts/lsig.algo.ts';

describe('Logic Signatures', function () {
  artifactsTest(PATH, 'BasicLsig', { lsig: true });
  artifactsTest(PATH, 'LsigWithArgs', { lsig: true });
  artifactsTest(PATH, 'LsigWithPrivateMethod', { lsig: true });

  describe('Compile Errors', function () {
    const errorMessages = {
      LsigInnerTxn: 'Inner transaction not allowed in logic signatures',
      LsigInnerLog: 'Logic signatures cannot log data',
      LsigMultipleMethods: 'Only one method called "logic" can be defined in a logic signature',
      LsigNonVoid: 'logic method must have a void return type',
    };

    Object.keys(errorMessages).forEach(function (contractName) {
      test(contractName, async function () {
        let msg: string;
        try {
          await compileAndCreate(
            algosdk.generateAccount(),
            'tests/contracts/lsig_compile_errors.algo.ts',

            contractName
          );
          msg = 'No error';
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
        } catch (e: any) {
          msg = e.message;
        }

        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        expect(msg).toMatch((errorMessages as any)[contractName]);
      });
    });
  });
});
