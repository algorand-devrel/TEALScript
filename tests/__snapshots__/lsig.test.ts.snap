// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Logic Signatures BasicLsig Artifacts Generates TEAL 1`] = `
"#pragma version 10
//#pragma mode logicsig
intcblock 1

// https://github.com/algorandfoundation/TEALScript

// The address of this logic signature is XBBFEDSZ6KEUC76ZFV5LUYHAYYRKAXRJ7FYUSBWIBTNA4HUYLMHGKGYXPU

b *route_logic

// logic()void
*route_logic:
	// execute logic()void
	callsub logic
	intc 0 // 1
	return

// logic(): void
logic:
	proto 0 0

	// tests/contracts/lsig.algo.ts:6
	// assert(true)
	intc 0 // 1
	assert
	retsub"
`;

exports[`Logic Signatures BasicLsig Artifacts Maintains program size 1`] = `18`;

exports[`Logic Signatures LsigWithArgs Artifacts Generates TEAL 1`] = `
"#pragma version 10
//#pragma mode logicsig
intcblock 1

// https://github.com/algorandfoundation/TEALScript

// The address of this logic signature is W6DDEFBKZB2HZYJ6WYJ3S3RSK57IRNV4LACOOLBGVDEWJ524RQWF5EBMTU

b *route_logic

// logic(uint64,uint64[])void
*route_logic:
	// b: uint64[]
	intc 0 // 1
	args
	extract 2 0

	// a: uint64
	pushint 0
	args
	btoi

	// execute logic(uint64,uint64[])void
	callsub logic
	intc 0 // 1
	return

// logic(a: AssetID, b: uint64[]): void
logic:
	proto 2 0

	// tests/contracts/lsig.algo.ts:13
	// assert(a)
	frame_dig -1 // a: AssetID
	assert

	// tests/contracts/lsig.algo.ts:14
	// assert(b[2])
	frame_dig -2 // b: uint64[]
	store 255 // full array
	load 255 // full array
	extract 16 8
	btoi
	assert
	retsub"
`;

exports[`Logic Signatures LsigWithArgs Artifacts Maintains program size 1`] = `39`;

exports[`Logic Signatures LsigWithPrivateMethod Artifacts Generates TEAL 1`] = `
"#pragma version 10
//#pragma mode logicsig
intcblock 1

// https://github.com/algorandfoundation/TEALScript

// The address of this logic signature is 2G7X6ZCAOFJOVGTOAGVL2ZKKJSLUSV2XFG4SKH3JRUIZ7EY7JWY3T7EQGI

b *route_logic

// privateMethod(): boolean
privateMethod:
	proto 0 1

	// tests/contracts/lsig.algo.ts:21
	// return true;
	intc 0 // 1
	retsub

// logic()void
*route_logic:
	// execute logic()void
	callsub logic
	intc 0 // 1
	return

// logic(): void
logic:
	proto 0 0

	// tests/contracts/lsig.algo.ts:25
	// assert(this.privateMethod())
	callsub privateMethod
	assert
	retsub"
`;

exports[`Logic Signatures LsigWithPrivateMethod Artifacts Maintains program size 1`] = `25`;
